### PR TITLE
R: Full support of lists and structs in R

### LINKED ISSUE #2841: R list column support (R is freezing)
Hi folks,

Every time I try to write an R dataframe with a column of 'list' type to a DuckDB database, R completely freezes with this message: `Error in duckdb_register(conn, view_name, value): Unsupported column type scan`.

Here is a MWE:
```
.path = tempfile()
con = DBI::dbConnect(duckdb::duckdb(), dbdir = .path, read_only = FALSE)

DBI::dbExecute(con, "CREATE TABLE df(id1 INTEGER not null,
                                 id2 VARCHAR not NULL, 
                                 int_list INT[], 
                                 varchar_list VARCHAR[],
                                 primary key (id1, id2))")

df <- tibble::tibble(id1 = as.integer(2), id2 = c("a"),
                     int_list = list(as.integer(1:3)),
                     varchar_list = list(c("a","b")))

DBI::dbWriteTable(con, 'df', df, append=TRUE)
```

(1) Can't DuckDB nested types `integer list column` and `varchar list column` handle R lists?
If not, is there a plan to support it or at least avoid crashing/freezing R?

(2) As a side-question, is there a way to alter DuckDB database to add Primary Keys? This would be important because the current method (declaring primary keys while creating a new table) requires a lot of manual work, like typing all columns of my R dataframe I want to copy to DuckDB (in some cases I have thousands of columns, which would make that impracticable).

Thanks a lot.

#### Environment (please complete the following information):
 - OS: Windows 10 Pro
 - DuckDB Version: 0.3.1-1
 - DuckDB Client: R

### PR DESCRIPTION
**Pull request by krlmlr**
This adds support for lists of atomic vectors. All elements must be of the same type for this to work. The commits make sense individually, but mostly depend on each other.

ALTREP support can be built on top of that.

Closes #2841.

### CODE REVIEW

File: tools/rpkg/src/scan.cpp [hunk 1]
```diff
@@ -6,20 +6,262 @@
...
+	auto child_rtype = rtype.GetListChildType();
+	for (idx_t i = 0; i < count; i++) {
+		auto val = source_data[i];
+		if (RSexpType::IsNull(val)) {
+			result_mask.SetInvalid(i);
+		} else {
+			auto result_data = FlatVector::GetData<list_entry_t>(result);
+			auto len = RApiTypes::GetVecSize(child_rtype, val);
+			result_data[i].offset = ListVector::GetListSize(result);
+			for (R_len_t i = 0; i < len; i++) {
```
Comment (Tmonster) [inline tools/rpkg/src/scan.cpp:97]: Just for my understanding, we are just pushing back the nested lists, but not every element in the list?

File: tools/rpkg/src/include/typesr.hpp [hunk 2]
```diff
@@ -39,11 +39,82 @@ enum class RType {
...
+
+	static RType LIST(const RType &child);
+	RType GetListChildType() const;
+
+	static RType STRUCT(child_list_t<RType> &&children);
+	child_list_t<RType> GetStructChildTypes() const;
+
+private:
+	RTypeId id_;
+	child_list_t<RType> aux_;
```
Comment (Tmonster) [inline tools/rpkg/src/include/typesr.hpp:110]: what does aux_ stand for in this case? Just auxiliary?

File: tools/rpkg/src/scan.cpp [hunk 3]
```diff
@@ -6,20 +6,262 @@
...
+		break;
+	}
+	case RType::DATE_INTEGER: {
+		auto data_ptr = (int *)coldata_ptr;
+		AppendColumnSegment<int, date_t, RDateType>(data_ptr, sexp_offset, v, this_count);
+		break;
+	}
+	case RType::LIST_OF_NULLS:
+	case RType::BLOB: {
+		auto data_ptr = (SEXP *)coldata_ptr;
```
Comment (Tmonster) [inline tools/rpkg/src/scan.cpp:246]: `BLOB` and `LIST_OF_NULLS` had different logic here https://github.com/duckdb/duckdb/pull/8503/files#diff-37ff3203dc9d0866d479ae910101108e08a2bdcf1f93bcf3f357c3be6adc94b3L371 . 
Why are they merged together now?

**Review by Tmonster — COMMENTED** [review]
Looks good to me. Would just like to see more test cases  for other types. 

If I'm not mistaken, nested types are registered per every nested child object?

File: tools/rpkg/src/include/typesr.hpp (hunk 2 above)
Comment (krlmlr) [reply tools/rpkg/src/include/typesr.hpp:110]: Yes, auxiliary information. Can you think of a better name?

File: tools/rpkg/src/scan.cpp (hunk 1 above)
Comment (krlmlr) [reply tools/rpkg/src/scan.cpp:97]: I'll rename the `i`. If this doesn't help, can you please rephrase?

File: tools/rpkg/tests/testthat/test_register.R [hunk 4]
```diff
@@ -72,8 +72,19 @@ test_that("experimental string handling works", {
...
+test_that("can register list of NULL values", {
+  con <- dbConnect(duckdb())
+  on.exit(dbDisconnect(con, shutdown = TRUE))
+  df <- tibble::tibble(a = 1:2, b = list(NULL))
+
+  expect_silent(duckdb_register(con, "df1", df))
+  expect_silent(duckdb_register(con, "df2", df[0, ]))
+
+  expect_equal(dbGetQuery(con, "SELECT * FROM df1"), as.data.frame(df))
+})
```
[the remark this answers is not in the document]
Comment (krlmlr) [reply tools/rpkg/tests/testthat/test_register.R:90]: BLOBs are tested excessively in DBItest. For this case, I added https://github.com/r-dbi/DBItest/issues/294.

**Comment by krlmlr** [comment]
> nested types are registered per every nested child object

Can you please rephrase?

**Comment by krlmlr** [comment]
I'll follow up with more tests.

**Comment by krlmlr** [comment]
I'm not too sure about my usage of `std::move()` .

**Review by Tmonster — COMMENTED** [review]
Was trying to test around and it seems like the following fails
```
df <- vctrs::data_frame(a = 1:4, b = list(4:6, 2:3, 1L, 5))
dbWriteTable(con, "df", df)
```
while
```
df <- vctrs::data_frame(a = 1:4, b = list(4:6, 2:3, 1L, 5L))
dbWriteTable(con, "df", df)
```
The error is `Error: rapi_execute: Can't convert R type`. I know that all nested structs/lists need the same type, but I think the user would appreciate feedback in this case that we don't yet support two types in the same nested object.

Might also be nice to have a RTypeToString() method for debugging? Possibly also for the above mentioned case?

Should we also test nested lists/structs that have above the default chunk size?

File: tools/rpkg/src/include/typesr.hpp (hunk 2 above)
Comment (Tmonster) [reply tools/rpkg/src/include/typesr.hpp:110]: maybe `child_Rtypes`? auxiliary feels to broad to me

File: tools/rpkg/src/scan.cpp (hunk 1 above)
Comment (Tmonster) [reply tools/rpkg/src/scan.cpp]: never mind, I think I figured it out

File: tools/rpkg/src/include/typesr.hpp (hunk 2 above)
Comment (krlmlr) [reply tools/rpkg/src/include/typesr.hpp:110]: `aux_` holds the list child type, struct child types and names, or factor levels. It's really a union type, but luckily everything fits into the `child_list_t` . Could be `extra_`, or perhaps `child_types_or_enum_levels_`, or a proper virtual base class with subclasses (but then how do we name that member?).

**Comment by Tmonster** [comment]
> I'm not too sure about my usage of `std::move()` .

Your usage of `move()` looks ok to me

**Comment by krlmlr** [comment]
Can you please construct a data frame that would exceed the chunk size limits?

Happy to tackle the detailed comments in a follow-up PR. The vctrs package in R also has a `list_of` type -- we can detect objects of this class (no need to drag in a dependency) to avoid scanning the entire list.

**Comment by Tmonster** [comment]
> Can you please construct a data frame that would exceed the chunk size limits?
Meant the STANDARD_VECTOR_SIZE, but it doesn't seem to be an issue. 

> Happy to tackle the detailed comments in a follow-up PR. The vctrs package in R also has a list_of type -- we can detect objects of this class (no need to drag in a dependency) to avoid scanning the entire list.

Yea, I think a follow up PR would be fine, this looks good to me.

**Review by Tmonster — APPROVED** [review]

**Comment by hannes** [comment]
Looks good to me, can we maybe now remove some of the excludes in test_types.R?

**Comment by krlmlr** [comment]
Absolutely. Can we do that in a follow-up PR?

**Comment by krlmlr** [comment]
(FWIW, `test-types.R` isn't affected by this PR. We can already output nested types today, this is about ingestion.)

**Pull request merged**